File: autopilot.py
# Flying Gator Autopilot Exploration
# (C) 2012 Cory Jones, Matthew Jadud, Tyler Khune

'''
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>. 
'''
# Imports
import socket
import re
import random
import math
from pid import PID

#pid responisble for taking in roll and changing ailerons
rollPID_ailerons = PID(3.0,0.04,0)
#PID responible for vertical speed
PID_VerticalSpeed = PID(3.0,0.04,0)
#PID responsible for controlling the pitch
PID_pitch = PID(0.38,0.0,0.0)
#PID responsible for controlling the altitude
PID_AltHold = PID(1.5,0.02,0.0)
PID_Heading = PID(0.1,0.0,0.0)
#holds the past altitude
pastAlt = 2000




# We'll do this once before flying.
def setup_plane():
  print "Running Setup."
  # Ailerons control roll.
  # We want to target a roll of zero.
  # So, we tell the rollPID to make its setPoint zero.
  global rollPID_ailerons
  global PID_heading
  global PID_pitch
  global PID_VerticalSpeed
  global PID_AltHold
  global PID_Heading
  global setHeading 
  setHeading = 40
  
  PID_AltHold.setPoint(200.00)
  rollPID_ailerons.setPoint(0)
  PID_VerticalSpeed.setPoint(-5)
  ##PID_pitch.setPoint(0)
  ##PID_AltHold.setPoint(0)
  PID_Heading.setPoint(setHeading)

# Here are the index values for each number that
# is in the "data" array.
'''
`((heading 0)
    (latitude-deg 1)
    (longitude-deg 2)
    (altitude-ft 3)
    (roll-deg 4)
    (pitch-deg 5)
    (heading-deg 6)
    (airspeed-kt 7)
    (aileron 8)
    (elevator 9)
    (rudder 10)
    (throttle 11)
'''
#this gets passed the object command
def control_plane (data):
  # Use the global PID structures
  global rollPID_ailerons
  global pastAlt
  
  #store the current pitch
  currentPitch = data[5]
  
  # Aileron, Elevator, Rudder, and Throttle
  cmd = [0,0,0,1]
  #cmd = data[8]
  cmd[1] = data[9]
	# We'll fly full throttle always.
  cmd[3] = 1.0
  # Do stuff and change how the plane flies.
  # Handle roll.
  # Get the current roll rate.
  roll = data[4]
  curAlt = data[3]
  # The roll has a range of -180 to 180.
  # Aileron control is from -1 to 1.
  # Divide the roll by 180 to get into aileron control range.
  scaled_roll = roll / 180.0
  # Now we need to update the PID controller with
  # the current roll data. This returns a value.
  # The value it returns is the value we want to use in controlling
  # our aircraft.
  #
  # By updating cmd.ailerons, we are ready to return the CMD
  # data structure and watch the aircraft fly.
  cmd[0] = (rollPID_ailerons.update(scaled_roll))
  #pitch controller
  cmd[1] = pitch_hold(currentPitch)
  
  #pass in the current altitude
  verticalSpeedHold(data[3])
  
  #pass in the current altitude
  altitudeHold(data[3])
  
  #set the past alt to the current alt before we exit
  pastAlt = data[3]
  current_heading = data[6]
  cmd[2] = heading(current_heading)
  
  return cmd

def heading(current_heading): 
  global setHeading
  setHeading = 180
  
  global PID_Heading
  global setHeading 
  
  PID_Heading.setPoint(setHeading)
  temp = 0
  temp = current_heading + 180
  
  #check to make sure temp is not greater than 360
  if temp <= 360:
    #if waypointHeading is greater than temp
    #you will need to turn left
    if current_heading > temp:
      #if current_heading minus 180 is less than zero,
      #then we need to still turn left
      if (current_heading - 180) <= 0:
          return -(PID_Heading.update(current_heading))
          
      else:
        return (PID_Heading.update(current_heading))
    #else the waypoint is less than current heading
    else:
      return (PID_Heading.update(current_heading))    
  else:
    return PID_Heading.update(current_heading)
    
#the pastalt holds the last value pof altitutde that the plane recorded
#alt is just a measurment passed in from the airplane telling us what the current altitude is. 
def rateofclimb(alt):
  #mult by 5 maybe?
  global pastAlt
  change_alt = alt - pastAlt
  return change_alt

#pastAlt is the past altitude
#the alt is the current altitude 
def altitudeHold(alt):
  global PID_AltHold
  global PID_VerticalSpeed
  goalAlt = 1000
  adjAlt = goalAlt - alt
  pidAlt = PID_AltHold.update(adjAlt)
  
  
  maxChange = 2
  
  if pidAlt >= maxChange:
    pidAlt = maxChange
  elif pidAlt < -(maxChange):
    pidAlt = -(maxChange)
  
  PID_VerticalSpeed.setPoint(pidAlt)
 
  
  
  
#alt is the current alt
#this edits the set point of the pidPitch controller
def verticalSpeedHold(alt):
  global PID_VerticalSpeed
  changeAlt = rateofclimb(alt)
  vertSpeed = PID_VerticalSpeed.update(changeAlt)
  PID_pitch.setPoint(vertSpeed)
  
  
  '''
  `((heading 0)
      (latitude-deg 1)
      (longitude-deg 2)
      (altitude-ft 3)
      (roll-deg 4)
      (pitch-deg 5)
      (heading-deg 6)
      (airspeed-kt 7)
      (aileron 8)
      (elevator 9)
      (rudder 10)
      (throttle 11)
  '''
  #working
def pitch_hold(curent_pitch):
  global PID_pitch
  curent_pitch = (-(curent_pitch))
  return PID_pitch.update(curent_pitch)
  
  
  
  
  

def main_loop():
	# UDP code from 
	# http://wiki.python.org/moin/UdpCommunication

	# The simulator is running on the same machine,
	# which is given the special loopback address
	# of '127.0.0.1'.
	SIM_IP = '127.0.0.1'

	# The simulator puts data out on port 5000, and 
	# reads data in on port 5010. These are parameters
	# that we give to the simulator when it starts, so
	# they could change, but only if we want them to.
	FROM_SIM = 5000
	TO_SIM   = 5010

	# Use the 'net, and send/read UDP packets.
	sim_sock = socket.socket (socket.AF_INET, socket.SOCK_DGRAM)

	# We have to "bind" to the port to listen for messages.
	# Note that we send a tuple: the IP address and the port.
	sim_sock.bind ((SIM_IP, FROM_SIM))

	while True:
		# recvfrom will block until data arrives.
		data, addr = sim_sock.recvfrom (1024)
		# pieces is now a list of strings.
		pieces = re.split (',', data)
		#numbers will be an array of strings.
		numbers = []
		for num in pieces:
			numbers.append(float(num))
    #print numbers
    # Set throttle to max and ailerons to something random
		# We send four pieces of data:
		# Aileron, Elevator, Rudder, and Throttle
		# IMPORTANT
		# Note how we construct a BYTE string by using the letter 'b' at the
		# start of the string. That matters a lot.
		print numbers
		CMD = control_plane(numbers)
		COMMAND = b"%f,%f,%f,%f\n" % (CMD[0], CMD[1], CMD[2], 1.0)
		print "[SEND] %s" % COMMAND
		sim_sock.sendto(COMMAND, (SIM_IP, TO_SIM))

# Run the setup
setup_plane()

# Run the main loop
main_loop()

File: autopilot.rb
require 'pid'
require 'socket.so'

# GLOBAL VARIABLE FOR UDP SOCKET
HOST_IP = "127.0.0.1"
PORT = 5000
TO_PORT = 5010

# This Creates a UDP socket which is used for sending
# and receiving data to the simulator
 
class UDPClient
  attr_accessor :host, :port, :socket, :to_port
  
  # Creates the object
  # @params 
  #   host: the IP of the host server
  #   port: the receiving port
  #   to_port: the port we are sending to
  
  def initialize(host, port, to_port)
    @host = host
    @port = port
    @to_port = to_port
    @socket = UDPSocket.new
  end
end

class ControlPanel
  attr_accessor :client, :heading, :latitude_deg, :longitude_deg, :roll_deg, :pitch_deg, :heading_deg, :airspeed_kt,
  :aileron, :elevator, :rudder, :throttle, :altitude_ft, :previous_alt
  
  def initialize
    socket_create()
    @pid_pitch = PID.new("Pitch", 3,1.25, 0.0, 0, 0, 1, -1)
    @rollpid_ailerons = PID.new("Roll", 3.0,0.04,0)
    # @pid_verticalspeed = PID.new(3.0,0.04,0)
    @pid_heading = PID.new("Heading", 0.05, 0, 0)
    @alt_error = 0
    @previous_alt = 0
    @desired_alt = 10000
    setup()
  end
  
  # Zeros out junk values
  def setup()
    @pid_pitch.setpoint(0)
    @rollpid_ailerons.setpoint(0)
    # @pid_verticalspeed.setpoint(-5)
    @pid_heading.setpoint(300)
  end
  
  def flipSign()
    if rand(100) > 50
      1
    else
      -1
    end
  end
  
  def noise(v, percentC, percentE)
    if rand(100) < percentC
      (v * rand(percentE)/100 * flipSign())
    else
      0.0
    end
  end
  
  # updates our control attributes
  # @params
  #   args: array with values of the aircraft's
  #         current settings
  def update(args)
    @heading = args[0].to_f
    @latitude_deg = args[1].to_f
    @longitude_deg = args[2].to_f
    @altitude_ft = args[3].to_f
    @roll_deg = args[4].to_f
    @pitch_deg = args[5].to_f
    @heading_deg = args[6].to_f
    @airspeed_kt = args[7].to_f
    @aileron = args[8].to_f
    @elevator = args[9].to_f
    @rudder = args[10].to_f
    @throttle = args[11].to_f
  end
  
  # Create a UDP socket and bind it
  
  def socket_create
    @client = UDPClient.new(HOST_IP, PORT, TO_PORT)
    @client.socket.bind(@client.host, @client.port)
  end
  
  # This function starts and runs everything
  # This is basically the autopilot
  # @params
  #   previous_run_time: time since the program as been
  #                      running
  def start(previous_run_time)
    start_time = previous_run_time
    # read in data from sim and update class  
      package = @client.socket.recvfrom(1024)
      array = package.to_s.split(',')
      update(array)
    # # # # # # #
    
    # divide the roll by 180 to get into aileron control range
    scaled_roll = (@roll_deg / 180.0)
    
    # Aileron, Elevator, Rudder, and Throttle
    control = { :aileron => @rollpid_ailerons.update(scaled_roll), # Now update the PID controller with the current roll data
                                                                    # This returns a value, the value returned is what we want 
                                                                    # to use to control the aircraft.
                :elevator => 0, 
                :rudder => 0, 
                :throttle => 2.0 }
    
    desired_rate = get_target_climb_rate
    current_rate = get_current_climb_rate(start_time)
    

    # TODO: Pitch needs to incorporate heading
    
    @pid_pitch.setpoint(desired_rate)
    control[:rudder] = heading()
    
    # We set the elevator to the negative of the pitch 
    # because of the simulator. + pitch make us go down,
    # - makes up go up
    control[:elevator] = -constrain("", @pid_pitch.update(current_rate), -0.2, 0.05)
    
    puts "CURRENT CLIMB::" + current_rate.to_s
    puts "DESIRED CLIMB::" + desired_rate.to_s
    puts "Altitude: " + @altitude_ft.to_s
    puts "Heading: " + @heading_deg.to_s
    
    # saving the current altitude before it updates
    @previous_alt = @altitude_ft
    return control
  end
  
  # this function returns the rudder control value from the PID, 
  # we have to do some math first before we call the PID controller.
  # def heading_old ()
  #    setHeading = 180   
  #    @pid_heading.setpoint(setHeading)
  #    temp = 0
  #    temp = @heading_deg + 180 
  #    if (temp <= 360)
  #      #if waypoint is greater than temp we need to go left.
  #      if ( (@heading_deg > temp) and (@heading_deg -180 <= 0) )           
  #        return -(@pid_heading.update(@heading_deg))
  #      end    
  #    end
  #    return(@pid_heading.update(@heading_deg))
  #  end
  
  def heading()
    @pid_heading.update(@heading_deg)
    
  end
  
  # Contrain a value between a min and a max
  
  def constrain (label, val, min, max)
    result = 0
    if val > max
      result = max
    elsif val < min
      result = min
    else
      result = val
    end
    puts label + result.to_s
    result
  end
    
  # Get the climb angle we need to have
  # based on the current altitude  
  def get_target_climb_rate
    @alt_error = (@desired_alt - @altitude_ft)
    constrain("climb", Math.atan2(@alt_error / 1000.0, 1.0), -0.08, 0.03)
  end
  
  # Gets the current climb rate
  # @params 
  #  start_time: this is the time when when whole iteration started
  def get_current_climb_rate( start_time )
    Math.atan2(( @altitude_ft - @previous_alt) / (Time.now - start_time)/1000.0, 1.0)
  end

end # end of ControlPanel


def time_diff_milli(start, finish)
   (finish - start) * 1000.0
end

def main
  control = ControlPanel.new
  end_time = Time.now
  
  while true
    sending = control.start(end_time)
    command = "%f,%f,%f,%f\n" % [sending[:aileron], sending[:elevator] , sending[:rudder] , sending[:throttle]]
    puts "[Send]" + command.to_s
    control.client.socket.send(command,0, control.client.host, control.client.to_port )
    end_time = Time.now
    #puts control.inspect
  end
end

main()